pkg/R/utils.R

```diff
@@ -288,3 +254,117 @@ convertEnvsToList <- function(keys, vals) {
            list(keys[[name]], vals[[name]])
          })
 }
+
+# Utility function to recursively traverse the Abstract Syntax Tree (AST) of a
+# user defined function (UDF), and to examine variables in the UDF to decide 
+# if their values should be included in the new function environment.
+# param
+#   node The current AST node in the traversal.
+#   oldEnv The original function environment.
+#   argNames A character vector of parameters of the function. Their values are
+#            passed in as arguments, and not included in the closure.
+#   newEnv A new function environment to store necessary function dependencies.
+processClosure <- function(node, oldEnv, argNames, newEnv) {
+  nodeLen <- length(node)
+  if (nodeLen == 0) {
+    return
+  }
+  if (nodeLen > 1 && typeof(node) == "language") {
+    # Recursive case: current AST node is an internal node, check for its children. 
+    if (length(node[[1]]) > 1) {
+      for (i in 1:nodeLen) {
+        processClosure(node[[i]], oldEnv, argNames, newEnv)
+      }
+    } else {  # if node[[1]] is length of 1, might be an R primitive.
+      nodeChar <- as.character(node[[1]])
```

Comment:
Just to make sure I understand this correctly, the reason we care about `<-` and `function` is that these are statements that change the values of argNames and defVar.  In that case I think we should also switch on `=` as that is also an assignment operator.

I am not sure I completely get why we need to special case `{`. Its just a symbol and should work fine using the default case ?